clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -0,0 +1,340 @@
+         */
+        private final List<TopicPartition> toForget;
+
+        /**
+         * All of the partitions which exist in the fetch request session.
+         */
+        private final Map<TopicPartition, PartitionData> sessionParts;
+
+        /**
+         * The metadata to use in this fetch request.
+         */
+        private final FetchMetadata metadata;
+
+        FetchRequestData(Map<TopicPartition, PartitionData> toSend,
+                         List<TopicPartition> toForget,
+                         Map<TopicPartition, PartitionData> sessionParts,
+                         FetchMetadata metadata) {
+            this.toSend = toSend;
+            this.toForget = toForget;
+            this.sessionParts = sessionParts;
+            this.metadata = metadata;
+        }
+
+        /**
+         * Get the set of partitions to send in this fetch request.
+         */
+        public Map<TopicPartition, PartitionData> toSend() {
+            return toSend;
+        }
+
+        /**
+         * Get a list of partitions to forget in this fetch request.
+         */
+        public List<TopicPartition> toForget() {
+            return toForget;
+        }
+
+        /**
+         * Get the full set of partitions involved in this fetch request.
+         */
+        public Map<TopicPartition, PartitionData> sessionParts() {
+            return sessionParts;
+        }
+
+        public FetchMetadata metadata() {
+            return metadata;
+        }
+    }
+
+    public class Builder {
+        /**
+         * The next partitions which we want to fetch.
+         *
+         * It is important to maintain the insertion order of this list by using a LinkedHashMap rather
+         * than a regular Map.
+         *
+         * One reason is that when dealing with FULL fetch requests, if there is not enough response
+         * space to return data from all partitions, the server will only return data from partitions
+         * early in this list.
+         *
+         * Another reason is because we make use of the list ordering to optimize the preparation of
+         * incremental fetch requests (see below).
+         */
+        private LinkedHashMap<TopicPartition, PartitionData> next = new LinkedHashMap<>();
+
+        /**
+         * Mark that we want to remove a partition from the incremental fetch session.
+         */
+        public void remove(TopicPartition topicPartition) {
+            next.remove(topicPartition);
+        }
+
+        /**
+         * Mark the data which we want from this partition on the upcoming fetch.
+         */
+        public void add(TopicPartition topicPartition, PartitionData data) {
+            next.put(topicPartition, data);
+        }
+
+        public FetchRequestData build() {
+            if (nextMetadata.isFull()) {
+                log.debug("Built full fetch {} for node {} with {} partition(s).",
+                    nextMetadata, node, next.size());
```

Comment:
Perhaps we may as well list out all the partitions?

Reply:
As in the comment above, though, there are going to be a huge number of them.  So it's not really practical (except maybe on TRACE level)